[mlir][linalg] Enable scalable vectorization of linalg.unpack

banach-space opened the pull request:
This patch updates `vectorizeAsTensorUnpackOp` to support scalable
vectorization by requiring user-specified vector sizes for the _read_ operation
(rather than the _write_ operation) in `linalg.unpack`. The rationale and an
example follow.

Conceptually, `linalg.unpack` consists of these high-level steps:
  * **Read** from the source tensor using `vector.transfer_read`.
  * **Transpose** the read value according to the permutation in the
    `linalg.unpack` op (via `vector.transpose`).
  * **Re-associate** dimensions of the transposed value, as specified by the op
    (via `vector.shape_cast`)
  * **Write** the result into the destination tensor via `vector.transfer_write`.

Previously, the vector sizes provided by the user were interpreted as
write-vector sizes. These were used to:
  * Infer read-vector sizes using the `inner_tiles` attribute of the unpack op.
  * Deduce vector sizes for the transpose and shape cast operations.
  * Ultimately determine the vector shape for the write.

However, this logic breaks when one or more tile sizes are dynamic. In such
cases, `vectorizeUnPackOpPrecondition` fails, and vectorization is rejected.

This patch switches the contract: users now directly specify the
"read-vector-sizes", which inherently encode all inner tile sizes - including
dynamic ones. It becomes the user's responsibility to provide valid sizes.

In practice, since `linalg.unpack` is typically constructed, tiled, and
vectorized by the same transformation pipeline, the necessary
"read-vector-sizes" should be recoverable.

With this change, we are finally able to vectorise `linalg.unpack` with
dynamic shapes using scalable vector sizes - see the included tests.

banach-space committed:
[mlir][linalg] Enable scalable vectorization of linalg.unpack (WIP)

This patch updates `vectorizeAsTensorUnpackOp` to support scalable
vectorization by requiring user-specified vector sizes for both the
_read_ and _write_ operations involved in `linalg.unpack`. Detailed
rationale and an example are provided below.

Conceptually, `linalg.unpack` consists of the following high-level steps:
  1. _Read_ from the source tensor.
  2. Transpose the value read in step (1).
  3. _Write_ the value from step (2) into the destination tensor.

Currently, when vectorizing with user-provided vector sizes, only the
sizes for the _write_ operation (step 3) are required. Sizes for the
_read_ operation (step 1) are inferred from static shapes and inner tile
sizes.

This logic breaks when the input shapes or tile sizes are dynamic
(indeed, `vectorizeUnPackOpPrecondition` rejects such cases ATM and the
vectorization fails). This patch addresses the issue by requiring
explicit vector sizes for both the read and write sides, enabling
scalable vectorization in such cases.

Example:

```mlir
func.func @unpack(%in: tensor<1x1x8x?xf32>, %out: tensor<8x?xf32>) -> tensor<8x?xf32> {
  %vs = vector.vscale
  %c8 = arith.constant 8 : index
  %tile_size = arith.muli %vs, %c8 : index

  %unpack = linalg.unpack  %in
    inner_dims_pos = [0, 1]
    inner_tiles = [8, %tile_size]
    into %out : tensor<1x1x8x?xf32> -> tensor<8x?xf32>
  return %unpack : tensor<8x?xf32>
}

module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%arg0: !transform.any_op {transform.readonly}) {
    %0 = transform.structured.match ops{["linalg.unpack"]} in %arg0 : (!transform.any_op) -> !transform.any_op
    transform.structured.vectorize %0 vector_sizes [1, 1, 8, [8],  8, [8]] : !transform.any_op
    //                                              \         /    \    /
    //                                              read-sizes   write-sizes
    transform.yield
  }
}
```

Finally, this patch also extends `createReadOrMaskedRead` and
`createWriteOrMaskedWrite` to take scalable flags.

egebeysel commented on mlir/test/Dialect/Linalg/vectorization/linalg-ops.mlir:
```diff
@@ -1173,7 +1252,7 @@ module attributes {transform.with_named_sequence} {
 func.func @test_vectorize_padded_pack(%arg0: tensor<32x7x15xf32>, %arg1: tensor<32x4x1x16x2xf32>) -> tensor<32x4x1x16x2xf32> {
   %pad = arith.constant 0.000000e+00 : f32
-  %pack = linalg.pack %arg0 padding_value(%pad : f32) inner_dims_pos = [2, 1] inner_tiles = [16, 2] into %arg1 : tensor<32x7x15xf32> -> tensor<32x4x1x16x2xf32>
+  %pack = linalg.pack %arg0 padding_value(%pad : f32) inner_dims_pos = [2, 1] inner_tiles = [16, [2]] into %arg1 : tensor<32x7x15xf32> -> tensor<32x4x1x16x2xf32>
```
why do we have this change?

egebeysel commented on mlir/test/Dialect/Linalg/vectorization/linalg-ops.mlir:
```diff
@@ -940,34 +940,113 @@ module attributes {transform.with_named_sequence} {
+module attributes {transform.with_named_sequence} {
+ transform.named_sequence @__transform_main(%arg0: !transform.any_op {transform.readonly}) {
+   %0 = transform.structured.match ops{["linalg.unpack"]} in %arg0 : (!transform.any_op) -> !transform.any_op
+   transform.structured.vectorize %0 vector_sizes [2, 1, [16], 2, 4, [16]] : !transform.any_op
```
Q: I guess this is "allowed" because we let the user specify the full read and write vector sizes, but do we basically allow any user-specified sizes now and assume it's UB of some sort if the specified sizes don't make sense?

egebeysel commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -2487,12 +2531,14 @@ vectorizeScalableVectorPrecondition(Operation *op,
   if (numOfScalableDims == 0)
     return success();
 
+  // TODO: Check the following!
```
Also here

egebeysel commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1860,25 +1866,54 @@ vectorizeAsTensorUnpackOp(RewriterBase &rewriter, linalg::UnPackOp unpackOp,
-  // vectorSizes is the shape of the vector that will be used to do final
+  // Split input-vector-sizes into vector sizes for the read and write
+  // operations.
+  if (!inputVectorSizes.empty()) {
```
One thing I'm curious about, even if we have fully static sizes and inner tiles, if the user specifies some size, we use that one. I think this was the previous behaviour as well, but should we maybe add some sort of sanity check for this? I'm not really familiar with the current behaviour either so this is rather a question on why things are the way they are :)

banach-space replied:
Thanks for these questions!

> even if we have fully static sizes and inner tiles, if the user specifies some size, we use that one

If a "user" provides the configuration, the compiler ought to respect that. Otherwise, we could have a situation where a user provides some input and that's unexpectedly ignored by the compiler. (*)

Also, if a "user" provides the configuration, it is similar to saying "Hey, compiler, I know better, follow my instructions! And ignore the static loop bounds that you may have access to - like I said - I know better." :)

Note that:
* If the user-specified vectors are too large, masking is used to make sure there are no out-of-bounds accesses.
* If the user-specified vectors are smaller than the actual run-time input then there won't be any out-of-bounds accesses anyway.

Hopefully this makes sense :)

> should we maybe add some sort of sanity check for this?

Yes, I have updated the pre-condition calculation in [this commit](https://github.com/llvm/llvm-project/pull/149293/commits/3b482fcb0013d971aa5befb9b99da166a34bb1a5). Can we do more and/or better? 🤔 

(*) One option would be for the compiler to issue some diagnostic saying that the user input was ignored. However, I personally feel that we should avoid such high-level complexities.

hanhanW replied:
Yes, we should verify if it has valid input vectors or not in precondition check, like the linalg one.

egebeysel reviewed (commented):
Thanks for this PR! I think the logic to extend the input vector sizes from write sizes to read & write sizes looks good, and to my understanding also makes sense to enable scalable vectorization. I've tested the WIP of this PR on downstream IREE and it seemed to serve what I was trying to do 😃  I'm not very familiar with the linalg vectorization, so I might've left more questions than comments, sorry about that! 🥲 Generally, I think more comments in code on what this change enables, what is the behaviour for user-specified sizes that are not "suitable" - if it's UB or some other future check fails etc. would help understand the behaviour better though.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -2073,6 +2113,9 @@ static LogicalResult
 vectorizeUnPackOpPrecondition(linalg::UnPackOp unpackOp,
                               ArrayRef<int64_t> inputVectorSizes) {
 
+  // FIXME!!!
```
Sorry, I was meant to remove that, it's no longer needed.

banach-space committed:
fixup! [mlir][linalg] Enable scalable vectorization of linalg.unpack (WIP)

Remove leftover code + comments

banach-space committed:
fixup! fixup! [mlir][linalg] Enable scalable vectorization of linalg.unpack (WIP)

Fix pre-condition calculation

banach-space committed:
fixup! fixup! [mlir][linalg] Enable scalable vectorization of linalg.unpack (WIP)

Improve documentation + fix test after rebasing on top of
* https://github.com/llvm/llvm-project/pull/150602

banach-space committed:
fixup! fixup! fixup! [mlir][linalg] Enable scalable vectorization of linalg.unpack (WIP)

Remove unintended test change

banach-space committed:
fixup! fixup! fixup! fixup! [mlir][linalg] Enable scalable vectorization of linalg.unpack (WIP)

Remove TODO

banach-space committed:
fixup! fixup! fixup! fixup! fixup! [mlir][linalg] Enable scalable vectorization of linalg.unpack (WIP)

Fix comment

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1902,17 +1937,20 @@ vectorizeAsTensorUnpackOp(RewriterBase &rewriter, linalg::UnPackOp unpackOp,
-  }
-  if (!outerDimsPerm.empty()) {
-    applyPermutationToVector(readVectorSizes, outerDimsPerm);
+  if (readVectorSizes.empty()) {
```
Thanks for noticing 🙏🏻 

I've rewritten this entirely, please take another look :)

banach-space commented:
>  I'm not very familiar with the linalg vectorization, so I might've left more questions than comments, sorry about that!

No need to apologize! You took the time to review my PR, and it’s totally expected that you'd have questions - I'm more than happy to clarify things :)

This code should be clear to everyone, not just those who've worked on it before, so thanks for highlighting areas that could use more explanation. I’m happy to expand further if anything’s still unclear.

I’ve pushed quite a few changes and added links to the relevant commits in the comments. That’s introduced a bit of noise, though, so it might be easier to review this as one large change rather than commit-by-commit.

Thanks again for taking a look 🙏🏻

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1866,59 +1879,76 @@ vectorizeAsTensorUnpackOp(RewriterBase &rewriter, linalg::UnPackOp unpackOp,
-  SmallVector<int64_t> vectorSizes(inputVectorSizes);
-  if (vectorSizes.empty()) {
-    llvm::append_range(vectorSizes, sourceShape.take_front(destSize));
+  // 1. Obtain vector sizes for the read and write operation.s
```
```suggestion
  // 1. Obtain vector sizes for the read and write operations.
```

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1866,59 +1879,76 @@ vectorizeAsTensorUnpackOp(RewriterBase &rewriter, linalg::UnPackOp unpackOp,
+      readVectorSizes[innerDimPos[index]] =
+          llvm::divideCeil(readVectorSizes[innerDimPos[index]], size);
+    }
+    if (!outerDimsPerm.empty()) {
+      applyPermutationToVector(readVectorSizes, outerDimsPerm);
+    }
```
style nit

```suggestion
    if (!outerDimsPerm.empty())
      applyPermutationToVector(readVectorSizes, outerDimsPerm);
```

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -2067,24 +2099,45 @@ vectorizeDynamicLinalgOpPrecondition(linalg::LinalgOp op,
-    LDBG("Inner-tiles must be constant: " << unpackOp << "\n");
+  // The input vector sizes must be equal to:
+  //  * read-vector-rank + write-vector-rank
+  if (!inputVectorSizes.empty()) {
+    if (inputVectorSizes.size() !=
+        unpackOp.getDestRank() + unpackOp.getSourceRank()) {
```
collapse two if into one if?

hanhanW replied:
bump, I think the comment is not addressed?

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -2488,10 +2542,11 @@ vectorizeScalableVectorPrecondition(Operation *op,
-  // non-linalg Ops so far
-  if (!linalgOp)
-    return failure();
+  // Cond 1: Reject Ops that don't implement the LinalgOp interface, with the
+  // exception of UnpackOp for which there is a dedicated hook.
+  if (!linalgOp) {
+    return isa<linalg::UnPackOp>(op) ? success() : failure();
+  }
```
Please also update the function comment. Now it accepts linalg.unpack ops.

Is the below form simpler to you?

```cpp
  if (!linalgOp)
    return success(isa<linalg::UnPackOp>(op));
```

https://github.com/llvm/llvm-project/blob/56ae79a6ab70ded92846e053b13745ea51e48089/llvm/include/llvm/Support/LogicalResult.h#L27-L31

banach-space replied:
> Is the below form simpler to you?

Yes, thanks for the suggestion 🙏🏻

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -2587,7 +2642,7 @@ vectorizeScalableVectorPrecondition(Operation *op,
                  isa<linalg::DepthwiseConv1DNwcWcOp>(op) ||
                  isa<linalg::MatvecOp>(op) || isa<linalg::Mmt4DOp>(op) ||
-                 hasReductionIterator(linalgOp));
+                 isa<linalg::UnPackOp>(op) || hasReductionIterator(linalgOp));
```
Why do we need the change? I think we already check it above, so it is a deadcode?

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1866,59 +1879,76 @@ vectorizeAsTensorUnpackOp(RewriterBase &rewriter, linalg::UnPackOp unpackOp,
-  // - if outer_dims_perms is present: do that permutation on readVectorSizes.
-  // - Append the remaining shape from SS
-  // E.g. let's say let's say unpackTensorType.getShape() = <8x8x32x16>
-  // inner Dim Pos = [0, 1] and Inner Tiles = [32, 16], vector_sizes are [512,
-  // 128] and outer_dims_perm is [1, 0] then read shape is:
-  //   ReadVectorSizes(initial): [512, 128]
-  //   Final Value(after innerDim Adjustment): [512/32, 128/16]
-  //                                           = [16, 8]
-  //   After applying outer_dims_perm: [8, 16]
-  //   After appending the rest of the sourceShape: [8, 16, 32, 16]
-
-  SmallVector<int64_t> readVectorSizes(vectorSizes.begin(), vectorSizes.end());
-
-  for (auto [index, size] : enumerate(innerTiles)) {
-    readVectorSizes[innerDimPos[index]] =
-        llvm::divideCeil(readVectorSizes[innerDimPos[index]], size);
-  }
-  if (!outerDimsPerm.empty()) {
-    applyPermutationToVector(readVectorSizes, outerDimsPerm);
+  // 1.2 Infer vector sizes for the read operation.
+  //
+  // The steps are:
+  //  1. readVectorSizes = writeVectorSizes
+  //  2. Take readVectorSizes from 1. and divide all locations pointed by
+  //     the inner_dims_pos attribyte by the `inner_tiles` attribute value.
+  //  3. If outer_dims_perms is present, permutate readVectorSizes accordingly.
+  //  4. Append the remaining sizes from the source tensor.
+  //
+  // Note, this will only work when all sizes are static!
+  if (readVectorSizes.empty()) {
+    readVectorSizes = writeVectorSizes;
+    for (auto [index, size] : enumerate(innerTiles)) {
+      readVectorSizes[innerDimPos[index]] =
+          llvm::divideCeil(readVectorSizes[innerDimPos[index]], size);
+    }
+    if (!outerDimsPerm.empty()) {
+      applyPermutationToVector(readVectorSizes, outerDimsPerm);
+    }
+    readVectorSizes.append(sourceShape.begin() + writeVectorSizes.size(),
+                           sourceShape.end());
```
IIUC, it happens when the inputVectorSize is not provided; it works only if the op has static shape. Then why can't it just take the shape from source and dest directly? E.g.,

```cpp
if (!inputVectorSizes.empty()) {
  ...
} else {
  auto destShape = unpackOp.getDestType().getShape();
  writeVectorSizes.assign(destShape.begin(), destShape.end());
  // same for readVectorSizes but it uses Source.
  useInBoundsInsteadOfMasking = true;
}
```

I think the inference was needed because we used to pass vector input sizes for dest shape?

hanhanW reviewed (changes requested):
I was thinking the same issue, i.e., require input vector size for both read and write, when I worked on pad op vectorization. I was thinking to provide two vector list, but appending one to the other one looks okay to me. Because it is op's implementation details, like tile sizes in tiling implementation.

Thanks for the patch, and I'll spend some time to integrate this into IREE: https://github.com/iree-org/iree/blob/a260a5e4c3033ed2aa35498865b856e68340b7dc/compiler/src/iree/compiler/Codegen/Utils/Utils.cpp#L1784C15-L1822

> This patch addresses the issue by requiring explicit vector sizes for both the read and write sides, enabling scalable vectorization in such cases.

We should document it somewhere. I don't find a good place, but maybe in `vectorizeUnPackOpPrecondition`'s function comment? I think it is better than nothing.

banach-space committed:
Simplify code as per comments from HanHan

egebeysel commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1831,126 +1832,138 @@ vectorizeAsTensorPackOp(RewriterBase &rewriter, linalg::PackOp packOp,
+  // writeVectorSizesFinal had to match the shapecast shape for dynamic sizes,
   // otherwise the validator complains that the mask size is invalid.
-  SmallVector<int64_t> writeVectorSizes(
+  // FIXME: We should not override write-vector-sizes like this.
+  SmallVector<int64_t> writeVectorSizesFinal(
```
I guess the fixme is not necessary anymore. Also, does the `writeVectorSizesFinal` implicitly do some verification?

banach-space replied:
In fact, we don't need `writeVectorSizes`, see https://github.com/llvm/llvm-project/pull/151325. Shortly I will post a note explaining "why".

egebeysel reviewed (approved):
Thanks for the changes! LGTM :)

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1831,126 +1832,138 @@ vectorizeAsTensorPackOp(RewriterBase &rewriter, linalg::PackOp packOp,
+///   // Reshape the data based on the target
+///   %sc = vector.shape_cast %tr : vector<1x8x1x8xf32> to vector<8x8xf32>
+///   // Write the result vector to the destination tensor.
+///   vector.transfer_write %sc into %dest : vector<8x8xf32>, tensor<8x8xf32>
```
optional nit: technically, it has a return value.
```suggestion
///   %write = vector.transfer_write %sc into %dest
        : vector<8x8xf32>, tensor<8x8xf32>
```

banach-space replied:
Thanks!

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1831,126 +1832,138 @@ vectorizeAsTensorPackOp(RewriterBase &rewriter, linalg::PackOp packOp,
-                         sourceShape.end());
 
-  Location loc = unpackOp->getLoc();
+  // CASE 1. 2: Vector sizes have to be inferred.
+  if (writeVectorSizes.empty()) {
+    if (ShapedType::isDynamicShape(destShape) ||
+        ShapedType::isDynamicShape(sourceShape))
+      return failure();
+
+    readVectorSizes.assign(sourceShape.begin(), sourceShape.end());
+    writeVectorSizes.assign(destShape.begin(), destShape.end());
+    useInBoundsInsteadOfMasking = true;
+  }
```
The comment needs to be updated.

Question: why not put them to the `else` branch?

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1831,126 +1832,138 @@ vectorizeAsTensorPackOp(RewriterBase &rewriter, linalg::PackOp packOp,
+
+  // CASE 1.1: Vector sizes are user-specified.
+  if (!inputVectorSizes.empty()) {
+    readVectorSizes.append(inputVectorSizes.begin(),
+                           inputVectorSizes.begin() + sourceShape.size());
+    writeVectorSizes.append(inputVectorSizes.begin() + sourceShape.size(),
+                            inputVectorSizes.end());
+    readScalableVectorFlags.append(inputScalableVecDims.begin(),
+                                   inputScalableVecDims.begin() +
+                                       sourceShape.size());
+    writeScalableVectorFlags.append(inputScalableVecDims.begin() +
+                                        sourceShape.size(),
+                                    inputScalableVecDims.end());
```
optional nit: I feel that `assign` is slightly better than `append`, because we won't need to check if they are empty or not when we browsed the code.

hanhanW reviewed (approved):
LGTM, just a few nits. Thanks!

banach-space committed:
Address the remaining comments from HanHan

banach-space commented:
**UPDATE: 30/7/25**

* This [commit](https://github.com/llvm/llvm-project/pull/149293/commits/56108b1df69e150c475adc58880ca7dce5355b21) addresses the remaining comments from @hanhanW . 
* I have rebased this PR on top of https://github.com/llvm/llvm-project/pull/151334. This rebase addresses this [comment](https://github.com/llvm/llvm-project/pull/149293#discussion_r2237499014) from @egebeysel .

**GENERAL OBSERVATIONS + FUTURE STEPS**

Having implemented #151334, I now realise that we don't require separate vector sizes for the _write_ operation (there's a small twist though).

To illustrate, take this example:
```mlir
func.func @example(%source: tensor<8x4x16x16xf32>, %dest: tensor<64x127xf32>) -> tensor<64x127xf32> {
   %0 = linalg.unpack %source outer_dims_perm = [1, 0] inner_dims_pos = [0, 1] inner_tiles = [16, 16] into %dest : tensor<8x4x16x16xf32> -> tensor<64x127xf32>
   return %0 : tensor<64x127xf32>
 }
```

It will be vectorized as:
```mlir
  func.func @example(%arg0: tensor<8x4x16x16xf32>, %arg1: tensor<64x127xf32>) -> tensor<64x127xf32> {
    %cst = arith.constant 0.000000e+00 : f32
    %c0 = arith.constant 0 : index
    // This is key - vec Op 1 !!!
    %0 = vector.transfer_read %arg0[%c0, %c0, %c0, %c0], %cst {in_bounds = [true, true, true, true]} : tensor<8x4x16x16xf32>, vector<8x4x16x16xf32>
    // This is key - vec Op 2 !!!
    %1 = vector.transpose %0, [1, 2, 0, 3] : vector<8x4x16x16xf32> to vector<4x16x8x16xf32>
    // This is key - vec Op 3 !!!
    %2 = vector.shape_cast %1 : vector<4x16x8x16xf32> to vector<64x128xf32>
    %c0_0 = arith.constant 0 : index
    // This is key - vec Op 4!!!
    %3 = vector.transfer_write %2, %arg1[%c0_0, %c0_0] {in_bounds = [true, false]} : vector<64x128xf32>, tensor<64x127xf32>
    return %3 : tensor<64x127xf32>
  }
```

Now, once we vectorize the read operation, the remaining sizes are already pre-determined (i.e. the sizes for the _write_ operation):
* For `vector.transpose`, the sizes must match the sizes from `vector.transfer_read` (% permutation).
* For `vector.shape_cast`, the input must match the output of `vector.transpose`. The output is uniquely determined by e.g. applying `outer_dims_perm` from `linalg.unpack` to the output from `vector.transpose`.
* For `vector.transfer_write`, we have to use the output shape from `vector.shape_cast`.

TL;Dr We should only require sizes for the _write_ operation.

**TWIST**

While we should be able to infer the scalable flags, there is some logic still missing. This should not be a problem though.

**NEXT STEPS**

While we could land this as is (IREE integration looks fine: https://github.com/iree-org/iree/pull/21514, thanks @hanhanW ) and then iterate in-tree, it might be "healthier" if there's one self-contained change. 

Let me refine this and then integrate into IREE (to make sure that the integration works). Also, @hanhanW , lets sync offline and make sure that switching to "only vector sizes for the read Op" is going to work for IREE.

WDYT?

banach-space committed:
Simplify to only require read-vector-sizes.

hanhanW reviewed (approved):
LGTM, thanks! Just 3 nits.

banach-space committed:
Final tweaks

egebeysel commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -2499,8 +2492,12 @@ vectorizePadOpPrecondition(tensor::PadOp padOp,
+///
+/// For Ops implementing the LinalgOp interface, this is quite restrictive - it
+/// models the fact that in practice we would only make selected dimensions
+/// scalable. For other Ops (e.g. `linalg.unpack`), this will succed
```
nit:
```suggestion
/// scalable. For other Ops (e.g. `linalg.unpack`), this will succeed
```

egebeysel reviewed (approved):
LGTM

banach-space committed:
Fix typo
